pkg/graveler/ref/manager.go

```diff
@@ -163,16 +165,61 @@ func (m *KVManager) ListRepositories(ctx context.Context) (graveler.RepositoryIt
 	return NewKVRepositoryIterator(ctx, &m.kvStore)
 }
 
+func (m *KVManager) updateRepoState(ctx context.Context, repo *graveler.RepositoryRecord, state graveler.RepositoryState) error {
+	repo.State = state
+	return m.kvStore.SetMsg(ctx, graveler.RepositoriesPartition(), []byte(graveler.RepoPath(repo.RepositoryID)), graveler.ProtoFromRepo(repo))
+}
+
+func (m *KVManager) deleteRepository(ctx context.Context, repo *graveler.RepositoryRecord) error {
+	// ctx := context.Background() TODO (niro): When running this async create a new context and remove ctx from signature
+	logger := logging.FromContext(ctx)
+	// Scan through repository partition and delete entities, using SecondaryIndex struct since we don't really need the data, just the key
+	// Failures on deleting repository entities don't constitute as repository deletion failure
+	partition := graveler.RepoPartition(repo)
+	it, err := kv.NewPartitionIterator(ctx, m.kvStore.Store, (&kv.SecondaryIndex{}).ProtoReflect().Type(), partition)
+	if err != nil {
+		logger.WithField("repository", repo.RepositoryID).
+			WithError(err).
+			Error("Error scanning repository partition")
+	}
+	defer it.Close()
+	for it.Next() {
+		v := it.Entry()
+		err = m.kvStore.DeleteMsg(ctx, partition, v.Key)
+		if err != nil && !errors.Is(err, kv.ErrNotFound) {
+			logger.WithField("repository", repo.RepositoryID).
+				WithField("key", string(v.Key)).
+				Error("Failed to delete repository entity")
+		}
+	}
+	if it.Err() != nil {
+		if err != nil {
+			logger.WithField("repository", repo.RepositoryID).
+				WithError(it.Err()).
+				Error("Error during iteration")
+		}
+	}
+
+	// Finally delete the repository record itself
+	return m.kvStore.DeleteMsg(ctx, graveler.RepositoriesPartition(), []byte(graveler.RepoPath(repo.RepositoryID)))
+}
+
 func (m *KVManager) DeleteRepository(ctx context.Context, repositoryID graveler.RepositoryID) error {
-	// TODO: delete me
-	// temp code to align with DB manager. Delete once https://github.com/treeverse/lakeFS/issues/3640 is done
-	_, err := m.getRepository(ctx, repositoryID)
+	repo, err := m.getRepositoryRec(ctx, repositoryID)
 	if err != nil {
 		return err
 	}
-	// END TODO: delete me
 
-	return m.kvStore.DeleteMsg(ctx, graveler.RepositoriesPartition(), []byte(graveler.RepoPath(repositoryID)))
+	// Set repository state to deleted and then perform background delete.
+	if repo.State != graveler.RepositoryState_IN_DELETION {
+		err = m.updateRepoState(ctx, repo, graveler.RepositoryState_IN_DELETION)
```

Comment:
I can think of a scenario where one thread deletes the state and then another one creates it with state `RepositoryState_IN_DELETION`. I think `updateRepoState` should use `SetIf`

Reply:
Unless I'm missing something - repo is always created in ACTIVE and the only scenario of updateRepoState currently is for delete. In this scenario I will always want to set it to deleted regardless of previous state. I suggest we defer this for when we will actually see the need for setIf. I'm afraid this will just introduce uninteresting bugs

Reply:
The `Delete` below might happen before some other flow updates the repo state to `RepositoryState_IN_DELETION`. Then you returned to that state, although there's nothing more to do.